Change the URL building in HttpHookAsync to match the behavior of HttpHook

alejorodriguez96 opened the pull request:
As mentioned in #37664 `HttpHookAsync` asumes that `base_url` is passed to the hook or can be extracted from connection, whereas `HttpHook` handles correctly when no `base_url` is passed and `endpoint` is a complete url. 

This PR aims to set the same behavior to both hooks.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

alejorodriguez96 committed:
fix http aysnc hook url if endpoint is not defined

Lee-W commented on airflow/providers/http/hooks/http.py:
```diff
@@ -38,6 +38,14 @@
 
+def _url_from_endpoint(base_url: str | None, endpoint: str | None) -> str:
+    """Combine base url with endpoint."""
+    print(f"base_url: {base_url}, endpoint: {endpoint}")
```
If we're logging stuffs, we might want to use `import logging` instead

alejorodriguez96 replied:
sorry debugging print left on code, fixed!

Lee-W commented on airflow/providers/http/hooks/http.py:
```diff
@@ -38,6 +38,14 @@
+    """Combine base url with endpoint."""
+    print(f"base_url: {base_url}, endpoint: {endpoint}")
+    if base_url and not base_url.endswith("/") and endpoint and not endpoint.startswith("/"):
+        return base_url + "/" + endpoint
```
```suggestion
        return f"{base_url}/{endpoint}"
```

alejorodriguez96 replied:
done! ty

sunank200 committed:
Introducing Logical Operators for dataset conditional logic (#37101)

* Implement | and & operators so that they can be used instead of DatasetAll and DatasetAny

* Refactor dataset class inheritance (#37590)

* Refactor DatasetAll and DatasetAny inheritance

They are moved from airflow.models.datasets to airflow.datasets since
the intention is to use them with Dataset, not DatasetModel. It is more
natural for users to import from the latter module instead.

A new (abstract) base class is added for the two classes, plus the OG
Dataset class, to inherit from. This allows us to replace a few
isinstance checks with simple molymorphism and make the logic a bit
simpler.

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
Co-authored-by: Wei Lee <weilee.rx@gmail.com>
Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

sunank200 committed:
Introducing Logical Operators for dataset conditional logic (#37101)

* Implement | and & operators so that they can be used instead of DatasetAll and DatasetAny

* Refactor dataset class inheritance (#37590)

* Refactor DatasetAll and DatasetAny inheritance

They are moved from airflow.models.datasets to airflow.datasets since
the intention is to use them with Dataset, not DatasetModel. It is more
natural for users to import from the latter module instead.

A new (abstract) base class is added for the two classes, plus the OG
Dataset class, to inherit from. This allows us to replace a few
isinstance checks with simple molymorphism and make the logic a bit
simpler.

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
Co-authored-by: Wei Lee <weilee.rx@gmail.com>
Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

Lee-W commented on tests/providers/http/hooks/test_http.py:
```diff
@@ -648,3 +648,25 @@ def test_process_extra_options_from_connection(self):
+            hook = HttpAsyncHook()
+            with mock.patch("aiohttp.ClientSession.post", new_callable=mock.AsyncMock) as mocked_function:
+                await hook.run("v1/test")
+                assert mocked_function.call_args.args[0] == schema + "://" + conn.host + "v1/test"
```
```suggestion
                assert mocked_function.call_args.args[0] == f"{schema}://{conn.host}v1/test"
```

alejorodriguez96 replied:
done! ty

Lee-W reviewed (commented):
Left a few nitpicks, but overall looks good!

uranusjr committed:
Implement AIP-60 Dataset URI formats (#37005)

uranusjr committed:
Implement AIP-60 Dataset URI formats (#37005)

alejorodriguez96 committed:
refactor some nits and rm print

alejorodriguez96 committed:
refactor some nits and rm print

alejorodriguez96 committed:
Revert "refactor some nits and rm print"

This reverts commit 5fa8aed5405feb778580de7eae5f32279b3c07e4.

dirrao reviewed (approved):
Nice work. LGTM.
